src/msg/Dispatcher.h

```diff
@@ -209,12 +209,14 @@ class Dispatcher {
    *
    * Do not acquire locks in this method! It is considered "fast" delivery.
    *
-   * return 1 for success
-   * return 0 for no action (let another Dispatcher handle it)
-   * return <0 for failure (failure to parse caps, for instance)
+   * Note: MonClient is the only caller of this method and it is configured
+   *       to only call a single dispatcher.
+   *
+   * return true for success (auth succeeds for this stage of session construction)
+   * return false for failure (failure to parse caps, for instance)
    */
-  virtual int ms_handle_fast_authentication(Connection *con) {
-    return 0;
+  virtual bool ms_handle_fast_authentication(Connection *con) {
+    return true;
```

Comment:
Any reason for not going pure virtual?

Reply:
Dispatchers that don't handle authentication would need to have the method which seemed unnecessarily obtrusive.

Reply:
I share the @idryomov's concern.

How about introducing an intermediary stage (with the default implementation) in the inheritance hierarchy to address those auth-uninterested dispatchers explicitly?

```
$ grep -r ms_handle_fast_authentication ./
./mon/MonClient.cc:      handle_authentication_dispatcher->ms_handle_fast_authentication(con);
./mon/MonClient.cc:    handle_authentication_dispatcher->ms_handle_fast_authentication(con);
./mon/Monitor.h:  int ms_handle_fast_authentication(Connection *con) override;
./mon/AuthMonitor.cc:	  mon.ms_handle_fast_authentication(s->con.get()) > 0) {
./mon/Monitor.cc:      ms_handle_fast_authentication(con);
./mon/Monitor.cc:    ms_handle_fast_authentication(con);
./mon/Monitor.cc:int Monitor::ms_handle_fast_authentication(Connection *con)
./msg/Dispatcher.h:  virtual int ms_handle_fast_authentication(Connection *con) {
./mds/MDSDaemon.h:  int ms_handle_fast_authentication(Connection *con) override;
./mds/MDSDaemon.cc:int MDSDaemon::ms_handle_fast_authentication(Connection *con)
./mgr/DaemonServer.h:  int ms_handle_fast_authentication(Connection *con) override;
./mgr/DaemonServer.cc:int DaemonServer::ms_handle_fast_authentication(Connection *con)
./osd/OSD.h:    int ms_handle_fast_authentication(Connection *con) override {
./osd/OSD.h:  int ms_handle_fast_authentication(Connection *con) override;
./osd/OSD.cc:int OSD::ms_handle_fast_authentication(Connection *con)
./test/fio/fio_ceph_messenger.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/perf_msgr_client.cc:    int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/perf_msgr_server.cc:  int ms_handle_fast_authentication(Connection *con) override {
```

```
$ grep -r public\ Dispatcher ./
./mon/Monitor.h:class Monitor : public Dispatcher,
./mon/MonClient.h:struct MonClientPinger : public Dispatcher,
./mon/MonClient.h:class MonClient : public Dispatcher,
./neorados/RADOSImpl.h:class RADOS : public Dispatcher
./tools/cephfs_mirror/ClusterWatcher.h:class ClusterWatcher : public Dispatcher {
./tools/cephfs/MDSUtility.h:class MDSUtility : public Dispatcher {
./librados/RadosClient.h:class librados::RadosClient : public Dispatcher,
./librbd/io/ImageDispatcher.h:class ImageDispatcher : public Dispatcher<ImageCtxT, ImageDispatcherInterface> {
./librbd/io/ObjectDispatcherInterface.h:  : public DispatcherInterface<ObjectDispatchInterface> {
./librbd/io/ObjectDispatcher.h:  : public Dispatcher<ImageCtxT, ObjectDispatcherInterface> {
./librbd/io/ImageDispatcherInterface.h:  : public DispatcherInterface<ImageDispatchInterface> {
./osdc/Objecter.h:class Objecter : public md_config_obs_t, public Dispatcher {
./mds/MDSDaemon.h:class MDSDaemon : public Dispatcher {
./mds/Beacon.h:class Beacon : public Dispatcher
./mds/MetricsHandler.h:class MetricsHandler : public Dispatcher {
./mds/MetricAggregator.h:class MetricAggregator : public Dispatcher {
./mgr/MgrStandby.h:class MgrStandby : public Dispatcher,
./mgr/DaemonServer.h:class DaemonServer : public Dispatcher, public md_config_obs_t
./mgr/MgrClient.h:class MgrClient : public Dispatcher
./osd/OSD.h:class OSD : public Dispatcher,
./osd/OSD.h:  struct HeartbeatDispatcher : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverSuite : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverTest : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverSuitePeer : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverTestPeer : public Dispatcher {
./test/crimson/test_messenger_peer.cc:class FailoverSuitePeer : public Dispatcher {
./test/crimson/test_messenger_peer.cc:class FailoverTestPeer : public Dispatcher {
./test/fio/fio_ceph_messenger.cc:class FioDispatcher : public Dispatcher {
./test/mon/test_mon_workloadgen.cc:class TestStub : public Dispatcher
./test/mon/test-mon-msg.cc:class MonClientHelper : public Dispatcher
./test/direct_messenger/test_direct_messenger.cc:class MockDispatcher : public Dispatcher {
./test/testmsgr.cc:class Admin : public Dispatcher {
./test/msgr/perf_msgr_client.cc:  class ClientDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class FakeDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class SyntheticDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class MarkdownDispatcher : public Dispatcher {
./test/msgr/perf_msgr_server.cc:class ServerDispatcher : public Dispatcher {
./client/Client.h:class Client : public Dispatcher, public md_config_obs_t {
```

For many of them the burden could as low as running `sed`.

Reply:
I think my concern in

https://github.com/ceph/ceph/pull/52939#discussion_r1367077257

was not actually an issue. So I've changed the default return to false. That should render this conversation moot.

Reply:
Well, I do need to adjust some daemons. Hmm.